### PR TITLE
feat(init): authorization 

### PR DESCRIPTION
**Pull request by RiezvykhDima**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,7 +4,8 @@
   "description": "```     https://hapi.dev/family/joi/ ```",
   "main": "index.js",
   "scripts": {
-    "test": "echo \"Error: no test specified\" && exit 1"
+    "build": "nodemon ./src/server/index.js",
```
Comment (v-checha) [inline package.json:7]: better use name 'dev' or 'start:dev'

File: src/components/Auth/index.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+        const emailIsExist = await UserService.isExistsEmail(value.email);
+        const nickNameIsExist = await UserService.isExists_nickName(value.nickName);
+
+        if (emailIsExist === true) throw new Error('email is taken by another user!'); 
+        if (nickNameIsExist === true) throw new Error('nickName is taken by another user!');
+
+        const hashedPassword = bcrypt.hashSync(value.password, HASH_SALT);
+
+        const userData = {
```
Comment (v-checha) [inline src/components/Auth/index.js:22]: better move business logic to service

File: src/components/Auth/index.js [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        const { error, value } = AuthValidation.signUp(req.body);
+        if (error) throw new ValidationError(error.details);
+
+        const emailIsExist = await UserService.isExistsEmail(value.email);
+        const nickNameIsExist = await UserService.isExists_nickName(value.nickName);
+
+        if (emailIsExist === true) throw new Error('email is taken by another user!'); 
+        if (nickNameIsExist === true) throw new Error('nickName is taken by another user!');
+
+        const hashedPassword = bcrypt.hashSync(value.password, HASH_SALT);
```
Comment (v-checha) [inline src/components/Auth/index.js:20]: try create hashed password with mongoose hooks pre 'save'
https://mongoosejs.com/docs/middleware.html#pre

File: src/components/Auth/index.js [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        });  
+    } catch (error) {
+        if (error instanceof ValidationError) {
+            return res.status(422).json({
+                message: error.name,
+                details: error.message,
+            });
+        }
+
+        res.status(500).json({
```
Comment (v-checha) [inline src/components/Auth/index.js:43]: response server error must be in error middleware

File: src/components/Auth/index.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+async function logIn(req, res, next) {
+    try {
+        const { error, value } = AuthValidation.logIn(req.body);
+        if(error) throw new ValidationError(error.details);
+
+        const existingUser = await UserService.findBy_nickName(value.nickName);
+        
+        if (existingUser === null) throw new Error('User does not exists!');
+
+        if (!bcrypt.compareSync(value.password, existingUser.password)) {
```
Comment (v-checha) [inline src/components/Auth/index.js:61]: move to auth service

File: src/components/Auth/index.js [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+        const tokens = AuthService.generateTokens(existingUser);
+
+        await AuthService.updateOrSaveToken(existingUser.id, tokens.refreshToken);
+
+
+        return res
+          .status(200)
+          .json({
+            tokens,
```
Comment (v-checha) [inline src/components/Auth/index.js:73]: all responses must contains field data, like 
`.json({ data: ...})`

File: src/components/Auth/index.js [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+          });
+    } catch(error) {
+        if(error instanceof ValidationError) {
+            return res.status(422).json({
+                message: error.name,
+                details: error.message,
+            });
+        }
+
+        res.status(500).json({
```
Comment (v-checha) [inline src/components/Auth/index.js:83]: the same

File: src/components/Auth/index.js [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        const user = await UserService.findById(requestToken.userId);
+
+        const tokens = generateTokens(user);
+
+        AuthService.updateOrSaveToken(user._id, tokens.refreshToken);
+
+        res
+          .status(200)
+          .json({
+            message: `accessToken Bearer ${tokens.accessToken}`
```
Comment (v-checha) [inline src/components/Auth/index.js:111]: when token is refreshed you must send to client new pare of tokens { access, refresh }

File: src/components/Auth/index.js [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    });
+} catch (error) {
+    if (error instanceof ValidationError) {
+        return res.status(422).json({
+            message: error.name,
+            details: error.message,
+        });
+    } 
+
+    res.status(500).json({
```
Comment (v-checha) [inline src/components/Auth/index.js:121]: the same

File: src/components/Auth/router.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
+const AuthComponent = require('.');
+const { Router } = require('express');
+
+const router = Router();
+
+router.post('/sign-up', AuthComponent.signUp);
```
Comment (v-checha) [inline src/components/Auth/router.js:6]: where route refresh?

File: src/components/User/index.js [hunk 11]
```diff
@@ -100,6 +100,36 @@ async function create(req, res, next) {
...
+        });
+    } catch (error) {
+        if (error instanceof ValidationError) {
+            return res.status(422).json({
+                message: error.name,
+                details: error.message,
+            });
+        }
+
+        res.status(500).json({
```
Comment (v-checha) [inline src/components/User/index.js:124]: the same

**Review by v-checha — COMMENTED** [review]
did not find middlewares

File: package.json (hunk 1 above)
Comment (RiezvykhDima) [reply package.json:7]: got it

File: src/components/Auth/index.js (hunk 2 above)
Comment (RiezvykhDima) [reply src/components/Auth/index.js:22]: got it

**Pull request closed without merging**